avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java

```diff
@@ -2132,26 +2133,107 @@ public static Schema valueSchema(Schema schema) {
     }
   }
 
-  private static boolean schemaEquals(Schema src, Schema that) {
+  private static boolean fieldListEquals(List<Field> one, List<Field> two,
+                                         Map<SchemaPair, Boolean> cache) {
+    if (one == two) {
+      return true;
+    } else if (one == null || two == null) {
+      return false;
+    } else {
+      ListIterator<Field> itOne = one.listIterator();
+      ListIterator<Field> itTwo = two.listIterator();
+      while (itOne.hasNext() && itTwo.hasNext()) {
+        if (!fieldEquals(itOne.next(), itTwo.next(), cache)) {
+          return false;
+        }
+      }
+      return itOne.hasNext() == itTwo.hasNext();
+    }
+  }
+
+  private static boolean fieldEquals(Field one, Field two, Map<SchemaPair, Boolean> cache) {
+    if (one == two) {
+      return true;
+    } else if (one == null || two == null) {
+      return false;
+    } else {
+      return one.getClass() == two.getClass()
+              && Objects.equals(one.index(), two.index())
+              && Objects.equals(one.name(), two.name())
+              && schemaEquals(one.schema(), two.schema(), cache);
+    }
+  }
+
+  private static class SchemaPair {
+    public Schema one;
+    public Schema two;
+
+    public SchemaPair(Schema one, Schema two) {
+      this.one = one;
+      this.two = two;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      SchemaPair that = (SchemaPair) o;
+      return Objects.equals(one, that.one)
+              && Objects.equals(two, that.two);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(one, two);
+    }
+  }
+
+  private static boolean schemaEquals(Schema src, Schema that, Map<SchemaPair, Boolean> cache) {
+    if (src == that) {
+      return true;
+    } else if (src == null || that == null) {
+      return false;
+    }
+
+    SchemaPair sp = new SchemaPair(src, that);
+    Boolean cacheHit = cache.get(sp);
+    if (cacheHit == null) {
+      // As long as we recurse into schemaEquals at the end of the method, we should be safe to
+      // default to true here.
```

Comment:
Maybe add a comment that we insert a temporary value of `true` here to handle cycles, and that this value will be overridden with the proper value at the end of this method.